src/main/java/eu/europeana/downloads/ListIdentifiersQuery.java

```diff
@@ -62,24 +77,77 @@ public String getVerbName() {
+            for (String setIdentifier : sets) {
+                 executeSingleThreadListRecord(oaipmhServer, setIdentifier);
             }
         }
     }
 
-    public List<String> getIdentifiers(OAIPMHServiceClient oaipmhServer) {
+    public List<String> getIdentifiers(OAIPMHServiceClient oaipmhServer, String setName) {
         List<String> identifiers = new ArrayList<>();
-        if (sets.isEmpty()) {
-            execute(oaipmhServer, null, identifiers);
-        } else {
-            for (String setName : sets) {
-                execute(oaipmhServer, setName, identifiers);
+        execute(oaipmhServer, setName, identifiers);
+        return identifiers;
+    }
+
+    private void executeSingleThreadListRecord(OAIPMHServiceClient oaipmhServer, String setIdentifier) {
+        List<String> identifiers = getIdentifiers(oaipmhServer, setIdentifier);
+        for(String identifier : identifiers) {
+            new GetRecordQuery(metadataPrefix, identifier, directoryLocation).execute(oaipmhServer);
+        }
+    }
+
+    private void executeMultithreadListRecords(OAIPMHServiceClient oaipmhServer, String setIdentifier) {
+        initThreadPool();
+
+        long counter = 0;
+        long start = System.currentTimeMillis();
+        ProgressLogger logger = new ProgressLogger(-1, logProgressInterval);
+
+        ListIdentifiersQuery identifiersQuery = prepareListIdentifiersQuery(setIdentifier);
+        List<String> identifiers = identifiersQuery.getIdentifiers(oaipmhServer, setIdentifier);
+        logger.setTotalItems(identifiers.size());
+
+        List<Future<ListRecordsResult>> results = null;
+        List<Callable<ListRecordsResult>> tasks = new ArrayList<>();
+
+        int perThread = identifiers.size() / threads;
+
+        // create task for each resource provider
+        for (int i = 0; i < threads; i++) {
+            int fromIndex = i * perThread;
+            int toIndex = (i + 1) * perThread;
+            if (i == threads - 1) {
+                toIndex = identifiers.size();
             }
+            tasks.add(new ListRecordsExecutor(identifiers.subList(fromIndex, toIndex), metadataPrefix, directoryLocation, oaipmhServer, logProgressInterval));
         }
-        return identifiers;
+
+        try {
+            // invoke a separate thread for each provider
+            results = threadPool.invokeAll(tasks);
+
+            ListRecordsResult listRecordsResult;
+            for (Future<ListRecordsResult> result : results) {
+                listRecordsResult = result.get();
+                LOG.info("Executor finished with {} errors in {} sec.",
+                        listRecordsResult.getErrors(), listRecordsResult.getTime());
+                counter += perThread;
+                logger.logProgress(counter);
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            LOG.error("Interrupted.", e);
+            Thread.currentThread().interrupt();
```

Comment:
Interrupt is called twice here. I noticed we do the same accidentally in ListRecordsQuery where this was probably copied from.